<?xml version="1.0"?>
<robot name="exoskeleton" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <xacro:property name="PI" value="3.1415926535897931"/>
  <!--waist property-->
  <xacro:property name="waist_mass" value="10" /> 
  <xacro:property name="waist_length" value="0.45" /> 
  <xacro:property name="waist_width" value="0.15" /> 
  <xacro:property name="waist_height" value="0.05" /> 
  <!--up property-->
  <xacro:property name="up_mass" value="8" /> 
  <xacro:property name="up_length" value="0.5" /> 
  <xacro:property name="up_width" value="0.1" /> 
  <xacro:property name="up_height" value="0.05" /> 
  <xacro:property name="up_effort" value="500" /> 
  <xacro:property name="up_lower" value="${-PI/2}" />
  <xacro:property name="up_upper" value="${2*PI/3}" />
  <xacro:property name="up_maxvel" value="${2*PI}" />
  <!--down property-->
  <xacro:property name="down_mass" value="3" /> 
  <xacro:property name="down_length" value="0.5" /> 
  <xacro:property name="down_width" value="0.1" /> 
  <xacro:property name="down_height" value="0.05" /> 
  <xacro:property name="down_effort" value="1500" /> 
  <xacro:property name="down_lower" value="0" /> 
  <xacro:property name="down_upper" value="${PI*2/3}" /> 
  <xacro:property name="down_maxvel" value="${PI*3}" /> 
  <!--foot property-->
  <xacro:property name="foot_back_mass" value="1.0" />
  <xacro:property name="foot_back_length" value="0.22" />
  <xacro:property name="foot_width" value="0.15" />
  <xacro:property name="foot_height" value="0.025" />
  <xacro:property name="foot_effort" value="500" /> 
  <xacro:property name="foot_front_lower" value="${-PI/3}" /> 
  <xacro:property name="foot_front_upper" value="0" /> 
  <xacro:property name="foot_back_lower" value="${-PI/18}" /> 
  <xacro:property name="foot_back_upper" value="${PI/6}" /> 
  <xacro:property name="foot_maxvel" value="${PI}" />

  <!--cam property-->
  <xacro:property name="camera_length" value="0.09"/>  <!-- 90 mm -->
  <xacro:property name="camera_width"  value="0.025"/> <!-- 25 mm -->
  <xacro:property name="camera_height" value="0.025"/> <!-- 25 mm -->
  <xacro:property name="camera_offset_x" value="0.0"/>
  <xacro:property name="camera_offset_y" value="0.0"/>
  <xacro:property name="camera_offset_z" value="0.0"/>
  <xacro:property name="camera_roll"  value="${PI/2}"/>  <!-- 90° -->
  <xacro:property name="camera_pitch" value="0.0"/>
  <xacro:property name="camera_yaw"   value="${PI/2}"/>  <!-- 90° -->
  <xacro:property name="camera_inclination" value="${PI/4}"/> <!-- 45° -->

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find exoskeleton)/robot_description/robot.gazebo" />
  <xacro:include filename="$(find exoskeleton)/robot_description/realsense_d435.xacro"/>


  <link name="waist_link">
    <collision>
      <origin xyz="${-waist_height/2-up_width/2} 0 0" rpy="0 0 0"/>
      <geometry>
	    <box size="${waist_height} ${waist_length} ${waist_width}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="${-waist_height/2-up_width/2} 0 0" rpy="0 0 0"/>
      <geometry>
	    <box size="${waist_height} ${waist_length} ${waist_width}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="${-waist_height/2-up_width/2} 0 0" rpy="0 0 0"/>
      <mass value="${waist_mass}"/>
      <inertia
	  ixx="${waist_mass / 12.0 * (waist_length*waist_length + waist_width*waist_width)}" ixy="0.0" ixz="0.0"
	  iyy="${waist_mass / 12.0 * (waist_height*waist_height + waist_width*waist_width)}" iyz="0.0"
	  izz="${waist_mass / 12.0 * (waist_height*waist_height + waist_length*waist_length)}"/>
    </inertial>
  </link>

    <joint name="camera_joint" type="fixed">
      <parent link="waist_link"/>
      <child link="camera_link"/>
      <origin xyz="${-up_width/2} 0 0" rpy="0 ${camera_inclination} 0"/>
    </joint>

  <link name="camera_link">
    <visual>
        <origin xyz="${camera_width} 0 0" rpy="${camera_roll} ${camera_pitch} ${camera_yaw}"/>
            <geometry>
                <mesh filename="$(find exoskeleton)/robot_description/realsense_d435.dae" scale="1 1 1"/>
            </geometry>
    </visual>
    <collision>
        <origin xyz="${camera_width} 0 0" rpy="${camera_roll} ${camera_pitch} ${camera_yaw}"/>
            <geometry>
                <mesh filename="$(find exoskeleton)/robot_description/realsense_d435.dae" scale="1 1 1"/>
            </geometry>
        </collision>
    <inertial>
        <mass value="0.1"/>
        <origin xyz="${camera_width} 0 0" rpy="${camera_roll} ${camera_pitch} ${camera_yaw}"/>
        <inertia ixx="1e-5" ixy="0" ixz="0" iyy="1e-5" iyz="0" izz="1e-5"/>
    </inertial>
  </link>

  <link name="d435_depth_optical_frame"/>
  <joint name="d435_depth_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="d435_depth_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 -${camera_inclination} 0"/>
  </joint>

  <link name="d435_color_optical_frame"/>
  <joint name="d435_color_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="d435_color_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 -${camera_inclination} 0"/>
  </joint>

  <link name="d435_infra1_optical_frame"/>
  <joint name="d435_infra1_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="d435_infra1_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 -${camera_inclination} 0"/>
  </joint>

  <link name="d435_infra2_optical_frame"/>
  <joint name="d435_infra2_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="d435_infra2_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 -${camera_inclination} 0"/>
  </joint>

  <joint name="LU_joint" type="revolute">
    <limit
        velocity="${up_maxvel}" effort="${up_effort}" lower="${up_lower}" upper="${up_upper}" />
    <parent link="waist_link"/>
    <child link="LU_link"/>
    <origin xyz="0 ${waist_length/2} 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="LU_link">
    <collision>
      <origin xyz="0 ${up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${up_width} ${up_height} ${up_length}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 ${up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${up_width} ${up_height} ${up_length}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="0 ${up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <mass value="${up_mass}"/>
      <inertia
	  ixx="${up_mass / 12.0 * (up_height*up_height + up_length*up_length)}" ixy="0.0" ixz="0.0"
	  iyy="${up_mass / 12.0 * (up_width*up_width + up_length*up_length)}" iyz="0.0"
	  izz="${up_mass / 12.0 * (up_width*up_width + up_height*up_height)}"/>
    </inertial>
  </link>

  <joint name="RU_joint" type="revolute">
    <limit
        velocity="${up_maxvel}"  effort="${up_effort}" lower="${up_lower}" upper="${up_upper}" />
    <parent link="waist_link"/>
    <child link="RU_link"/>
    <origin xyz="0 ${-(waist_length/2)} 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="RU_link">
    <collision>
      <origin xyz="0 ${-up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${up_width} ${up_height} ${up_length}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 ${-up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${up_width} ${up_height} ${up_length}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="0 ${-up_height/2} ${-(up_length/2)}" rpy="0 0 0"/>
      <mass value="${up_mass}"/>
      <inertia
	  ixx="${up_mass / 12.0 * (up_height*up_height + up_length*up_length)}" ixy="0.0" ixz="0.0"
	  iyy="${up_mass / 12.0 * (up_width*up_width + up_length*up_length)}" iyz="0.0"
	  izz="${up_mass / 12.0 * (up_width*up_width + up_height*up_height)}"/>
    </inertial>
  </link>

  <joint name="LD_joint" type="revolute">
    <limit
        velocity="${down_maxvel}"  effort="${down_effort}" lower="${down_lower}" upper="${down_upper}" />
    <parent link="LU_link"/>
    <child link="LD_link"/>
    <origin xyz="0 0 ${-up_length}" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="LD_link">
    <collision>
      <origin xyz="0 ${-down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${down_width} ${down_height} ${down_length}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 ${-down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${down_width} ${down_height} ${down_length}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="0 ${-down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <mass value="${down_mass}"/>
      <inertia
	  ixx="${down_mass / 12.0 * (down_height*down_height + down_length*down_length)}" ixy="0.0" ixz="0.0"
	  iyy="${down_mass / 12.0 * (down_width*down_width + down_length*down_length)}" iyz="0.0"
	  izz="${down_mass / 12.0 * (down_width*down_width + down_height*down_height)}"/>
    </inertial>
  </link>

  <joint name="RD_joint" type="revolute">
    <limit
       velocity="${down_maxvel}" effort="${down_effort}" lower="${down_lower}" upper="${down_upper}" />
    <parent link="RU_link"/>
    <child link="RD_link"/>
    <origin xyz="0 0 ${-up_length}" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="RD_link">
    <collision>
      <origin xyz="0 ${down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${down_width} ${down_height} ${down_length}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 ${down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <geometry>
	    <box size="${down_width} ${down_height} ${down_length}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="0 ${down_height/2} ${-(down_length/2)}" rpy="0 0 0"/>
      <mass value="${down_mass}"/>
      <inertia
	  ixx="${down_mass / 12.0 * (down_height*down_height + down_length*down_length)}" ixy="0.0" ixz="0.0"
	  iyy="${down_mass / 12.0 * (down_width*down_width + down_length*down_length)}" iyz="0.0"
	  izz="${down_mass / 12.0 * (down_width*down_width + down_height*down_height)}"/>
    </inertial>
  </link>

  <joint name="LF_joint" type="revolute">
    <limit
       velocity="${foot_maxvel}" effort="${foot_effort}" lower="${foot_back_lower}" upper="${foot_back_upper}" />
    <parent link="LD_link"/>
    <child link="LF_link"/>
    <origin xyz="-0.05 ${-down_height} ${-down_length}" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="LF_link">
    <gazebo>
        <self_collide>1</self_collide>
    </gazebo>
    <collision>
      <origin xyz="${foot_back_length/2-down_width/2} ${-foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <geometry>
	    <box size="${foot_back_length} ${foot_width} ${foot_height}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="${foot_back_length/2-down_width/2} ${-foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <geometry>
	    <box size="${foot_back_length} ${foot_width} ${foot_height}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="${foot_back_length/2-down_width/2} ${-foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <mass value="${foot_back_mass}"/>
      <inertia
	  ixx="${foot_back_mass / 12.0 * (foot_width*foot_width + foot_height*foot_height)}" ixy="0.0" ixz="0.0"
	  iyy="${foot_back_mass / 12.0 * (foot_back_length*foot_back_length + foot_height*foot_height)}" iyz="0.0"
	  izz="${foot_back_mass / 12.0 * (foot_back_length*foot_back_length + foot_width*foot_width)}"/>
    </inertial>
  </link>

  <joint name="RF_joint" type="revolute">
    <limit
       velocity="${foot_maxvel}" effort="${foot_effort}" lower="${foot_back_lower}" upper="${foot_back_upper}" />
    <parent link="RD_link"/>
    <child link="RF_link"/>
    <origin xyz="-0.05 ${down_height} ${-down_length}" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7"/>
  </joint>
  <link name="RF_link">
    <collision>
      <origin xyz="${foot_back_length/2-down_width/2} ${foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <geometry>
	    <box size="${foot_back_length} ${foot_width} ${foot_height}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="${foot_back_length/2-down_width/2} ${foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <geometry>
	    <box size="${foot_back_length} ${foot_width} ${foot_height}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <origin xyz="${foot_back_length/2-down_width/2} ${foot_width/2} ${-foot_height/2}" rpy="0 0 0"/>
      <mass value="${foot_back_mass}"/>
      <inertia
	  ixx="${foot_back_mass / 12.0 * (foot_width*foot_width + foot_height*foot_height)}" ixy="0.0" ixz="0.0"
	  iyy="${foot_back_mass / 12.0 * (foot_back_length*foot_back_length + foot_height*foot_height)}" iyz="0.0"
	  izz="${foot_back_mass / 12.0 * (foot_back_length*foot_back_length + foot_width*foot_width)}"/>
    </inertial>
  </link>

  <!--transmission staff for effort joint-->
  <transmission name="tran_LU">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="LU_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_LU">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_RU">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="RU_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_RU">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_LD">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="LD_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_LD">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_RD">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="RD_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_RD">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_LF">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="LF_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_LF">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_RF">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="RF_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_RF">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <xacro:gazebo_d435
     camera_name="d435"
     reference_link="camera_link"
     topics_ns="camera"
     depth_optical_frame="d435_depth_optical_frame"
     color_optical_frame="d435_color_optical_frame"
     infrared1_optical_frame="d435_infra1_optical_frame"
     infrared2_optical_frame="d435_infra2_optical_frame"
  />

</robot>
